### PR TITLE
Ball collision

### PR DESCRIPTION
**Pull request by naomijub**

### CODE REVIEW

File: src/ball/mod.rs [hunk 1]
```diff
@@ -65,12 +47,28 @@ impl Ball {
...
+    let paddle_points_distance = squared_dist(v, w);
+    let dot = ((p[0] - v[0]) * (w[0] - v[0]) + (p[1] - v[1]) * (w[1] - v[1])) / paddle_points_distance;
+    let g = 0f32.max(dot.min(1f32));
+    return squared_dist(p, na::Vector2::new( v[0] + g * (w[0] - v[0]), v[1] + g * (w[1] - v[1]) )).sqrt()
+  }
+
 
 #[test]
 fn update_moves_ball() {
+    let playes = &[&Player::new(-5f32, -5f32), &Player::new(-5f32, -5f32)];
```
Comment (evaporei) [inline src/ball/mod.rs]: Typo, `playes` instead of `players`.

File: src/ball/mod.rs [hunk 2]
```diff
@@ -21,38 +21,20 @@ impl Ball {
...
+
+        players.iter()
             .map(|p| p.paddle_vertex())
-            .for_each(|player_paddle| {
-                let left_upper_corner = player_paddle[0];
-                let left_bottom_corner = player_paddle[1];
-                let right_upper_corner = player_paddle[2];
-                let right_bottom_corner = player_paddle[3];
+            .for_each(|p| {
+                if p.chunks(2).any(|c|
```
Comment (evaporei) [inline src/ball/mod.rs]: Wow, the code here is very clean :slightly_smiling_face: 
I mean, I don't understand the mathematics, but it looks clean hahaha

File: src/ball/mod.rs [hunk 3]
```diff
@@ -21,38 +21,20 @@ impl Ball {
...
+                    dist_to_segment(ball_vec, c[0], c[1]) < 25f32
```
Comment (evaporei) [inline src/ball/mod.rs]: Maybe we could remove the magic number (25), and put it into a constant like `BALL_RADIUS`

File: src/ball/mod.rs [hunk 4]
```diff
@@ -21,38 +21,20 @@ impl Ball {
...
+                ) {
+                    self.velocity = -self.velocity;
```
Comment (evaporei) [inline src/ball/mod.rs]: I was thinking if we could put more meaning here, like creating a `change_direction` method, what do you think?

File: src/ball/mod.rs [hunk 5]
```diff
@@ -72,3 +68,55 @@ fn update_moves_ball() {
...
 }
+
+#[test]
+fn direction_changes_when_hitting_upper_wall() {
+    let mut ball = Ball {
+        position: na::Vector2::new(400f32, 26f32),
+        direction: na::Vector2::new(0f32, -1.1f32),
+    };
+
+    ball.rolling_stones();
```
Comment (evaporei) [inline src/ball/mod.rs:81]: The name rolling_stones was kept :joy:

File: src/main.rs [hunk 6]
```diff
@@ -5,6 +5,7 @@ use ggez::{ContextBuilder, GameResult};
 
 mod ball;
 mod draw;
+mod math;
```
Comment (evaporei) [inline src/main.rs:8]: Nice!

File: src/ball/mod.rs (hunk 4 above)
Comment (naomijub) [reply src/ball/mod.rs]: This is called direction now kkkkkk, but we could, maybe when we pair

File: src/ball/mod.rs (hunk 5 above)
Comment (naomijub) [reply src/ball/mod.rs:81]: yes <3

**Review by evaporei — APPROVED** [review]

**Pull request merged**